src/System.Linq/src/System/Linq/AppendPrepend.cs

```diff
@@ -0,0 +1,472 @@
+                        _state = 3;
+                        goto case 3;
+                    case 3:
+                        if (LoadFromEnumerator())
+                        {
+                            return true;
+                        }
+
+                        _state = 4;
+                        goto case 4;
+                    case 4:
+                        if (_index != _appCount)
+                        {
+                            _current = _appended[_index];
+                            ++_index;
+                            return true;
+                        }
+                        break;
+                }
+
+                Dispose();
+                return false;
+            }
+
+            public override AppendPrependIterator<TSource> Append(TSource element)
+            {
+                if (RunningOnCreatingThread && _appended.Append(element, _appCount))
+                {
+                    return new AppendPrependNIterator<TSource>(_source, _prepended, _preCount, _appended, _appCount + 1);
+                }
+
+                return new AppendPrepend1Iterator<TSource>(this, element, true);
+            }
+
+            public override AppendPrependIterator<TSource> Prepend(TSource element)
+            {
+                if (RunningOnCreatingThread)
+                {
+                    if (_prepended == null)
+                    {
+                        return new AppendPrependNIterator<TSource>(_source, new SharedCollection<TSource>(element), 1, _appended, _appCount);
+                    }
+
+                    if (_prepended.Prepend(element, _preCount))
+                    {
+                        return new AppendPrependNIterator<TSource>(_source, _prepended, _preCount + 1, _appended, _appCount);
+                    }
+                }
+
+                return new AppendPrepend1Iterator<TSource>(this, element, false);
+            }
+
+            public override int GetCount(bool onlyIfCheap)
+            {
+                IIListProvider<TSource> listProv = _source as IIListProvider<TSource>;
+                if (listProv != null)
+                {
+                    int count = listProv.GetCount(onlyIfCheap);
+                    return count == -1 ? -1 : count + _preCount + _appCount;
+                }
+
+                return !onlyIfCheap || _source is ICollection<TSource> || _source is ICollection ? _source.Count() + _preCount + _appCount : -1;
+            }
+
+            public override TSource[] ToArray()
+            {
+                int count = GetCount(onlyIfCheap: true);
+                if (count == -1)
+                {
+                    return EnumerableHelpers.ToArray(this);
+                }
+
+                TSource[] array = new TSource[count];
+                int index = 0;
+                for (int i = -_preCount; i != 0; ++i, ++index)
+                {
+                    array[index] = _prepended[i];
```

Comment:
Better still, perhaps `SharedCollection` should have a `CopyTo` to reduce indirection even further.